Bullet physics

CrosRoad95 opened the pull request:
Adds bullet physics engine to mta.

Plan changed.
Initial pr ( this pr ) will contain:

physicsCreateBoxShape, physicsCreateRigidBody, physicsCreateStaticCollision, physicsGetDebugLines + destroyElement, setElementPosition, setElementRotation support for rigid body and static collision

[tests v1.zip](https://github.com/multitheftauto/mtasa-blue/files/8459389/tests.v1.zip)

More details at mta development discord

<details>
<summary>More info</summary>
<p>
= SHERED =
physics-box-shape physicsCreateBoxShape( Dimensions size ) - creates box, cube shape. Minimum size of each dimension: 0.0001, maximum: 8196
physics-rigid-body physicsCreateRigidBody( physics-shape shape, RigidBodyOptions options ) - spawns a rigid body with given options, when spawned in air, it will immidently starts falling. Shape must be convex ( see below ). In future: eg.: some types of triangle mesh will not accepted, but compound-shape will even technically is not convex
physics-static-collision physicsCreateStaticCollision( physics-shape, Vector3 position, Vector3 rotation ) - spawns static collision, it will not affected by gravity, always stay in one place (like a platform)
table physicsGetDebugLines( Vector3 position, float radius ) - returns table {{fromX, fromY, fromZ, toX, toY, toZ, color}, ...} with collision wireframe at given position and radius. Send to client and loop to see current state of simulation. Color indicate current state of rigid body.

= CLIENT ONLY =
void physicsDrawDebug() - renders all debug lines, use in onClientRender

= Types =
RigidBodyOptions - table with accepted keys: {["mass"] = float, ["localInertia"] = Vector3, ["centerOfMass"] = Vector3, ["position"] = Vector3, ["rotation"] = Vector3}
Dimensions- can be x,y,z, or one number with all dimensions

= Hierarchy, type description:
physics-element - base of any physics element
physics-world-element : physics-element - element that has position and rotation
physics-shape : physics-element - base of all physics shape
physics-convex-shape : physics-shape - base of all convex shape
physics-box-shape : physics-convex-shape - box shape
physics-rigid-body : physics-world-element - shape affected by gravity ( eg: car, player, ball, some gta models )
physics-static-collision : physics-world-element - non moving collision ( eg: terrains, buildings, roads :D )

=== Extras:
physics-rigid-body and physics-static-collisions is always made of single physics-shape
treat physics-shape as a instance, can be used in multiple collisions and compound shape ( in future )

=== extended functionality of already existing functions:
set/getElementPosition, set/getElementRotation, destroyElement, now accept physics-world-element
getUserdataType - returns physics element type
set/getElementVelocity accept physics-rigid-body

=== **important:**
destroyElement side effect:
  due to hierarchy style of physics elements, if you destroy shape, all references ( rigid-bodies and static-collisions ) get destroyed too, see example below

=== nice to know:
- gta sa collision has precision 1/128m and maximum size 256x256x256, bp has float precision, and maximum size of float max number ( 3 + 38 zeros )
- spawning a lot of rigid bodies at same position may cause high ram usage ( eg spawn 1000 bodies at same time at same position ) - it can also crash due ran out of ram
- rigid bodies can sleep, it means when nothing hit them they will behave like static collision what increase performance


samples:

```lua
local box = physicsCreateBoxShape(1)
local rb1 = physicsCreateRigidBody(box, { position = {0,0,5} })
local rb2 = physicsCreateRigidBody(box, { position = {0,0,6} })
local rb3 = physicsCreateRigidBody(box, { position = {0,0,7} })

local floor = physicsCreateStaticCollision(box, 0,0,0)
destroyElement(box) -- rb1, rb2, rb3, floor and box get destroyed
```

</p>

Additional test script that tests rotation confirms that setElementRotation working exact the same for objects and bullet physics static collisions ( rigid body using same logic ):
```lua
local box = physicsCreateBoxShape(1)
local boxCollision1 = physicsCreateStaticCollision(box, 0,0,4)
local boxObject1 = createObject(1271, 0,0,4)
local boxCollision2 = physicsCreateStaticCollision(box, 0,3,4)
local boxObject2 = createObject(1271, 0,3,4)
local boxCollision3 = physicsCreateStaticCollision(box, 0,6,4)
local boxObject3 = createObject(1271, 0,6,4)

local i = 0;
addEventHandler("onClientRender", root, function()
  i = i + 0.5
  if(i > 360)then
    i = -360
  end
  physicsDrawDebug()
  setElementRotation(boxCollision1, i, 0,0)
  setElementRotation(boxObject1, i, 0,0)
  setElementRotation(boxCollision2, 0, i, 0)
  setElementRotation(boxObject2, 0, i, 0)
  setElementRotation(boxCollision3, 0,0, i)
  setElementRotation(boxObject3, 0,0, i)
  dxDrawText("Current rotation: "..i, 0,0, 100, 100)
end)


fadeCamera(true)
setCameraTarget(localPlayer)
```
![mta-screen_2022-11-24_19-50-18](https://user-images.githubusercontent.com/22455534/203850251-55d6672d-9163-4401-afaf-0d4c0259ce4f.png)


</details>

CrosRoad95 commented:
i'm waiting for quick, general review of code then i'll continue work on it

CrosRoad95 commented:
anyone has anything to say about this project? What should be added/changed to marge it?

qaisjp commented:
I've given your changes a very quick skim.

API stuff:

1. is it possible for some stuff that might take some time (and not required to be ready immediately) to be asynchronous? Such as [physicsCreateShapeFromModel](https://wiki.multitheftauto.com/wiki/physicsCreateShapeFromModel) and [physicsBuildCollisionFromGTA](https://wiki.multitheftauto.com/wiki/physicsBuildCollisionFromGTA)? I have not looked into whether this is even feasible.
2. Does [setElementPosition](https://wiki.multitheftauto.com/wiki/setElementPosition)/[Rotation](https://wiki.multitheftauto.com/wiki/setElementRotation) work on these physics objects?
    1. When you do [physicsAddChildShape](https://wiki.multitheftauto.com/wiki/physicsAddChildShape), is child.position @ (0,0,0) == parent.position?
    2. Would it be possible to use [setElementParent](https://wiki.multitheftauto.com/wiki/setElementParent), [setElementPosition](https://wiki.multitheftauto.com/wiki/setElementPosition), and [setElementRotation](https://wiki.multitheftauto.com/wiki/setElementRotation) here?
3. It looks like all the physics element kinds share the element type as `physics-element`. And that [physicsGetElementType](https://wiki.multitheftauto.com/wiki/PhysicsGetElementType) returns the 'inner type'. Not sure how I feel about this. I think [getElementType](https://wiki.multitheftauto.com/wiki/getElementType) should return the inner type, and people can use [physicsIsElement](https://wiki.multitheftauto.com/wiki/PhysicsIsElement) to determine if an element is a physics element (which is better than checking that the element type begins with `"physics-"`).
4. [physicsIsElement](https://wiki.multitheftauto.com/wiki/PhysicsIsElement) is weirdly named


However, for the code itself, here are some recommendations for _basic_ improvements:

1. use smart pointers
2. use `enum class` instead of `enum`
3. use the new Lua arg parser. you might need to extend the lua arg parser to have extra features. Send those as separate pull requests to master, so that those can be merged before this is merged.
4. Client/mods/deathmatch/logic/CClientPhysics.cpp imports `../Client/game_sa/CCameraSA.h`. Not allowed!

Generally:

1. I don't know if we're approving this feature. It's big and scary. We need to ask ourselves if this is in scope for MTA.

Pirulax commented:
IMHO, first fixing the other PR would be nice(the gtasa collision customization), because that also fixes setScale to  resize the collision which is neat.

CrosRoad95 commented:
Someone smart, we need idea how to sync this!!!!
even simple simulations may flood server with packets, batching will necessary.

I have few solution for it:
1. make option to create physics serverside too, and if something get created serverside it will synced.
2. make function like "setSyncId( element, number )" and if two or more players set same id, it will synced across that players. Server get function to set who is a syncer.

Depends on distance from element to rigid body, frequency of synchronization should change with option to change behaviour.

Commit:
optional heightmap for terrain

Commit:
addendum

Commit:
remove not needed headers

Commit:
improve api

Commit:
triangle mesh to bvh triangle mesh

Commit:
gimpact triangle mesh

Commit:
rename gimpact shape function name

Commit:
fix compilation error

test

Commit:
fix compilation error

test 2

Commit:
fix heigthfieldterrain size limits

Commit:
fix gimpact shape

Commit:
draw debug refactor, draw distance, use new parser

Commit:
std::monostate

Commit:
point to point constraint, constraints fixes

Commit:
fixes, cleanup for pointtopoint constraint

Commit:
fix disable collision between linked bodies

Commit:
remove unused enums, enum class for properties

Commit:
compound shape

Commit:
removing unmanaged pointers

Commit:
enum class for debug enum

Commit:
fix debugmode, fix warnings

Commit:
fix shared_ptr as input

Commit:
refactor

Commit:
Changing trianglemesh vertex position, update stack

Commit:
physicsSetHeight

Commit:
improve performance stats

Commit:
raycast

Commit:
efactor

Commit:
physicsSetEnable/IsEnabled

Commit:
compound shape fixes

Commit:
null to nullptr

Commit:
move common code to different file

Commit:
fix for get shape/rigids/staticollision/constraints functions

Commit:
raycast all details

Commit:
distance for ray casts

Commit:
raycasts

Commit:
raycast all upper limit and default limit


raycast all upper limit and default limit

Commit:
get position and rotation for staticcollision, shapecast

Commit:
shapecast and raycasts fixes

Commit:
clean up includes

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
CLuaPhysicsWorldElement class

Commit:
CLuaPHysicsWorldElement class

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
refactor

Commit:
refactor focused on multithreading

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
refactor, remove not needed includes

Commit:
getting shared elements refactor

Commit:
constraints, properties

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
refactor setting properties

Commit:
remove not needed code

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
update premake

Commit:
fix compilation

Commit:
refactor create function

Commit:
small refactor

Commit:
clean up, refactor

Commit:
move physics to shared folder

Commit:
fix compilation

Commit:
btVector3 <-> CVector conversion

Commit:
fix cvector, missing loadfunctions

Commit:
fix return cvector, fix crashes, dopulse serverside

Commit:
fix PhysicsGetElementType

Commit:
tomp remove constraints

Commit:
refactor

Commit:
refactor elo

Commit:
refactor properties

Commit:
use list instead of stack

Commit:
CLuaSharedFunctionParseHelpers

Commit:
remove functions responsible for copying collision to bullet

Commit:
fix compilation

Commit:
fix local inertia calculation

Commit:
attempt to fix compilation

Commit:
physicsGetDebugLines

Commit:
fix crash, fix names of userdata

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
fixes, refactor, crash fixes

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
fix leak, collect results

Commit:
flush changes when needed

Commit:
refactor, profiling for some functions

Commit:
refactor, function to test if is during simulation

Commit:
refactor

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
fixes, refactor

Commit:
options map for create function

Commit:
bulletphysics, thread support

Commit:
multithreaded, optimization

Commit:
better multithreading, island function, some fixes, new create world properties

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
getting island rigidbodies

Commit:
improve classes hierarchy

Commit:
refactor managers

Commit:
refactor, fix some crashes

Commit:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

Commit:
refactor setting properties

Commit:
refactor

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
format files

CrosRoad95 committed:
code improvments

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
refactor

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
add comma to premake5.lua

CrosRoad95 committed:
fixes

CrosRoad95 committed:
fixes

CrosRoad95 committed:
improve destryoing for future updates

CrosRoad95 committed:
rename CLuaPhysicsSharedLogic

CrosRoad95 committed:
improvments

CrosRoad95 committed:
improvments, WorldContext

CrosRoad95 committed:
improved profiler

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
refactor

CrosRoad95 committed:
fixes improvments, refactors

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
improve destroying

CrosRoad95 committed:
refactors, fixes

CrosRoad95 committed:
refactor

CrosRoad95 committed:
improve consistency of internal shape

CrosRoad95 committed:
refactor

CrosRoad95 committed:
fixes, refactors

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
fixes

CrosRoad95 committed:
fix crashes, general improvments

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
removing static collision reference when deleting collision

CrosRoad95 committed:
fix static collision memory leak

CrosRoad95 committed:
change "create" to "new"

based on v8 code base

CrosRoad95 committed:
update bulletphysics vendor to version 3.09

CrosRoad95 committed:
update vendor

required every vendor update
move stdinc.h

CrosRoad95 committed:
fixes, clearning references, compound shape

CrosRoad95 committed:
fix memory leaks, improve shape creation, fix crash when destroying compoun shape

CrosRoad95 committed:
refactor bounding box and sphere into separete functions

CrosRoad95 committed:
refactor getMargin

CrosRoad95 committed:
fix setEnable/IsEnable

CrosRoad95 committed:
isPhysicsElement

not it working

CrosRoad95 committed:
improved compound shape, added shape count function for compound

CrosRoad95 committed:
refactors, functions fixes

CrosRoad95 committed:
general improvments


temp fix

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
remove 98% of features

CrosRoad95 committed:
Merge branch 'BulletPhysics3D-With-More-Features' into BulletPhysics3D

CrosRoad95 committed:
remove 98% of features

CrosRoad95 committed:
remove old version of bp

CrosRoad95 committed:
fixes

CrosRoad95 committed:
remove atomic bool

CrosRoad95 committed:
refactor, remove features to re-add them in future, remove idea of "physicsCreateWorld"

CrosRoad95 committed:
fixes, refactors, remove not needed code

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
some fixes

CrosRoad95 committed:
everything

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
compilation fix

CrosRoad95 committed:
compilation fix 2

CrosRoad95 committed:
compilation fix 3

CrosRoad95 committed:
compilation fix 3

CrosRoad95 committed:
compilation fix 5

CrosRoad95 committed:
compilation fix 6

CrosRoad95 committed:
compilation fix 7

CrosRoad95 committed:
compilation fix 8

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
fixes

CrosRoad95 committed:
fixes

CrosRoad95 committed:
refactor, move physics functions to shared class

CrosRoad95 committed:
fixes, improvments, ect

CrosRoad95 committed:
fix get/set position

CrosRoad95 committed:
fix compilation

CrosRoad95 committed:
revert formatting

CrosRoad95 committed:
destroying physics elements

gta191977649 commented:
> I've given your changes a very quick skim.
> 
> API stuff:
> 
> 1. is it possible for some stuff that might take some time (and not required to be ready immediately) to be asynchronous? Such as [physicsCreateShapeFromModel](https://wiki.multitheftauto.com/wiki/physicsCreateShapeFromModel) and [physicsBuildCollisionFromGTA](https://wiki.multitheftauto.com/wiki/physicsBuildCollisionFromGTA)? I have not looked into whether this is even feasible.
> 2. Does [setElementPosition](https://wiki.multitheftauto.com/wiki/setElementPosition)/[Rotation](https://wiki.multitheftauto.com/wiki/setElementRotation) work on these physics objects?
>    
>    1. When you do [physicsAddChildShape](https://wiki.multitheftauto.com/wiki/physicsAddChildShape), is child.position @ (0,0,0) == parent.position?
>    2. Would it be possible to use [setElementParent](https://wiki.multitheftauto.com/wiki/setElementParent), [setElementPosition](https://wiki.multitheftauto.com/wiki/setElementPosition), and [setElementRotation](https://wiki.multitheftauto.com/wiki/setElementRotation) here?
> 3. It looks like all the physics element kinds share the element type as `physics-element`. And that [physicsGetElementType](https://wiki.multitheftauto.com/wiki/PhysicsGetElementType) returns the 'inner type'. Not sure how I feel about this. I think [getElementType](https://wiki.multitheftauto.com/wiki/getElementType) should return the inner type, and people can use [physicsIsElement](https://wiki.multitheftauto.com/wiki/PhysicsIsElement) to determine if an element is a physics element (which is better than checking that the element type begins with `"physics-"`).
> 4. [physicsIsElement](https://wiki.multitheftauto.com/wiki/PhysicsIsElement) is weirdly named
> 
> However, for the code itself, here are some recommendations for _basic_ improvements:
> 
> 1. use smart pointers
> 2. use `enum class` instead of `enum`
> 3. use the new Lua arg parser. you might need to extend the lua arg parser to have extra features. Send those as separate pull requests to master, so that those can be merged before this is merged.
> 4. Client/mods/deathmatch/logic/CClientPhysics.cpp imports `../Client/game_sa/CCameraSA.h`. Not allowed!
> 
> Generally:
> 
> 1. I don't know if we're approving this feature. It's big and scary. We need to ask ourselves if this is in scope for MTA.

Why it's "It's big and scary"? I think it opens a new door for the developer especially for those who wants to do some real simulations.

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
fix compilation

CrosRoad95 committed:
fix crash while restarting resource

CrosRoad95 committed:
fix draw debug

CrosRoad95 committed:
set seed to 0

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 commented:
waiting for review

tederis commented on Shared/sdk/CVector.h:
```diff
@@ -260,5 +262,9 @@ class CVector
+    inline bool operator!=(const CVector& param) const noexcept { return !(*this == param); }
+
+    std::tuple<float, float, float> AsXYZ() const { return {fX, fY, fZ}; }
+    CVector(const btVector3& other) : fX(other.x()), fY(other.y()), fZ(other.z()) {}
```
I don't think we should involve third party structures into MTA-utility classes. Rather, you can create a new header file that will contain functions for conversion between different types.

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsRigidBodyProxy.h:
```diff
@@ -0,0 +1,55 @@
+        SetEnabled(false);
+    }
+
+    static std::unique_ptr<CPhysicsRigidBodyProxy> New(CLuaPhysicsShape* pShape, const float fMass, CVector vecLocalInertia, CVector vecCenterOfMass,
```
We must avoid passing by value and prefer references instead

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsRigidBodyProxy.cpp:
```diff
@@ -0,0 +1,72 @@
+{
+    CPhysicsSharedLogic::SetPosition(pMotionstate->m_centerOfMassOffset, vecCenterOfMass);
+    btCollisionShape* pCollisionShape = pShape->InternalGetBtShape();
+    if (vecLocalInertia.LengthSquared() == 0)
```
Use an epsilon to compare two floating points

CrosRoad95 replied:
why not allow to pass very, very small vectors as a offset of local inertia?

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.h:
```diff
@@ -0,0 +1,57 @@
+    float GetLineWidth() const { return m_fLineWidth; }
+    void  SetDrawDistance(float fDistance) { m_fDrawDistance = fDistance; }
+    float GetDrawDistance() const { return m_fDrawDistance; }
+    void  SetCameraPosition(CVector vecPosition) { m_vecCameraPosition = vecPosition; }
```
Again, prefer to pass by a reference

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.h:
```diff
@@ -0,0 +1,57 @@
+    void drawLine(const btVector3& from, const btVector3& to, const btVector3& color);
+    void drawSphere(const btVector3& p, btScalar radius, const btVector3& color){};
+    void drawTriangle(const btVector3& a, const btVector3& b, const btVector3& c, const btVector3& color, btScalar alpha);
+    void drawContactPoint(const btVector3& PointOnB, const btVector3& normalOnB, btScalar distance, int lifeTime, const btVector3& color);
```
What notation are you using here?

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.h:
```diff
@@ -0,0 +1,57 @@
+    void Clear();
+
+    void DrawLine(const btVector3& from, const btVector3& to, const btVector3& lineColor);
+    void drawLine(const btVector3& from, const btVector3& to, const btVector3& fromColor, const btVector3& toColor);
```
Inconsistency in methods naming

CrosRoad95 replied:
methods with lower case are from interface, required by bp. "DrawLine" is actual draw line, maybe i should rename it to "DrawLineInternal"???

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.h:
```diff
@@ -0,0 +1,39 @@
+    static std::unique_ptr<btRigidBody> CreateRigidBody(btCollisionShape* pShape, const float fMass, const CVector vecLocalInertia,
+                                                        const CVector vecCenterOfMass);
+
+    static void EulerToQuaternion(const btVector3 rotation, btQuaternion& result);
```
Why are you avoiding references when passing arguments?

CrosRoad95 commented:
Could someone do a review? after that ill apply changes

Lpsd commented:
> Could someone do a review? after that ill apply changes

Maybe it's best to address the previous review first, before someone does another review, as they are all valid points.

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

# Conflicts:
#	Client/mods/deathmatch/premake5.lua

CrosRoad95 committed:
fixes along tederis suggestions

CrosRoad95 committed:
more  fixes along tederis suggestions

CrosRoad95 committed:
rename drawLine to drawLineInternal

CrosRoad95 commented:
> > Could someone do a review? after that ill apply changes
> 
> Maybe it's best to address the previous review first, before someone does another review, as they are all valid points.

ready

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
add missing velocity and rotation functions

CrosRoad95 committed:
client/server position/rotation/velocity/angularvelocity functions fix

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
fixes, mainly focused on destroying elements

CrosRoad95 committed:
fixes focused on destroying physics elements

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
creash fixes

CrosRoad95 committed:
temp fix, make debugGetLines return yellow lines

CrosRoad95 committed:
render debuglines only when requested

CrosRoad95 committed:
fixes, isPhysicsElement

CrosRoad95 committed:
on resource stop remove only physics element created by this resource

CrosRoad95 committed:
remove all bp files, replace them with download action

CrosRoad95 committed:
uncomment other actions

CrosRoad95 committed:
fix vendor gitignore

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
install bp on macos and linux

CrosRoad95 committed:
attempt to fix

CrosRoad95 committed:
attempt to fix linux macos compilation

CrosRoad95 committed:
Merge branch 'BulletPhysics3D' of https://github.com/CrosRoad95/mtasa-blue into BulletPhysics3D

CrosRoad95 committed:
attempt to fix

CrosRoad95 committed:
attempt to fix

CrosRoad95 committed:
attempt to fix

CrosRoad95 committed:
replace filter

CrosRoad95 committed:
attempt to fix

CrosRoad95 committed:
test, skip sha256 check

CrosRoad95 committed:
test install unzip

CrosRoad95 committed:
test

CrosRoad95 committed:
attempt fix zip

CrosRoad95 committed:
test fix

CrosRoad95 committed:
cleanup

CrosRoad95 committed:
maybe fix compilation

CrosRoad95 committed:
fix - getting position of physics-box-shape cause crashed

CrosRoad95 committed:
fix crash - inproper way of removing physics element

CrosRoad95 commented:
Waiting for review!
bp3 dependency got reworked and sources are no longer a part of this pull request!
Heavy testing is in progress, sources has been cleaned, unused code removed.

CrosRoad95 committed:
remove unused methods

CrosRoad95 committed:
cleanup

CrosRoad95 committed:
cleanup, unify names

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
cleanup

CrosRoad95 committed:
final cleanup

CrosRoad95 committed:
don't draw debug when game minimalized

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
improve getUserdataType

CrosRoad95 committed:
fix linux-build

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
remove old, unused createElement method

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

github-actions[bot] committed:
Update client en_US pot

tederis commented on Client/mods/deathmatch/logic/CClientGame.h:
```diff
@@ -278,6 +278,7 @@ class CClientGame
     CRPCFunctions*                GetRPCFunctions() { return m_pRPCFunctions; }
     CSingularFileDownloadManager* GetSingularFileDownloadManager() { return m_pSingularFileDownloadManager; };
     CServerInfo*                  GetServerInfo() { return m_ServerInfo.get(); }
+    CBulletPhysics*               GetPhysics() { return m_pPhysics.get(); }
```
Getters should be marked as constant

tederis commented on Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -38,6 +38,10 @@ static CClientExplosionManager*  m_pExplosionManager;
 static CClientProjectileManager* m_pProjectileManager;
 static CClientSoundManager*      m_pSoundManager;
 
+static CLuaPhysicsRigidBodyManager*       m_pLuaPhysicsRigidBodyManager;
```
Variables must always be initialized before use

CrosRoad95 replied:
is is, at 83-85 line in the same way as statics above

tederis commented on Server/mods/deathmatch/logic/CGame.h:
```diff
@@ -254,6 +255,7 @@ class CGame
     SharedUtil::CAsyncTaskScheduler* GetAsyncTaskScheduler() { return m_pAsyncTaskScheduler; }
 
     std::shared_ptr<CTrainTrackManager> GetTrainTrackManager() { return m_pTrainTrackManager; }
+    CBulletPhysics*                     GetPhysics() { return m_pPhysics.get(); }
```
Getters should be marked as constant

tederis commented on Shared/mods/deathmatch/logic/luadefs/CLuaPhysicsDefs.h:
```diff
@@ -0,0 +1,52 @@
+    static CLuaPhysicsRigidBody*       PhysicsCreateRigidBody(lua_State* luaVM, CLuaPhysicsShape* pPhysicsShape, CVector vecPosition,
+                                                              std::optional<CVector> vecRotation,
+                                                        RigidBodyOptions options);
+    static CLuaPhysicsStaticCollision* PhysicsCreateStaticCollision(lua_State* luaVM, CLuaPhysicsShape* pPhysicsShape, CVector position,
```
You should probably **CVector vecPosition** with **const CVector& vecPosition**

tederis commented on Shared/mods/deathmatch/logic/luadefs/CLuaPhysicsDefs.h:
```diff
@@ -0,0 +1,52 @@
+    static void            AddClass(lua_State* luaVM);
+
+    static bool                  IsPhysicsElement(lua_State* luaVM);
+    static CLuaPhysicsRigidBody*       PhysicsCreateRigidBody(lua_State* luaVM, CLuaPhysicsShape* pPhysicsShape, CVector vecPosition,
```
You should probably replace **CVector vecPosition** with **const CVector& vecPosition**

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsBaseManager.h:
```diff
@@ -0,0 +1,65 @@
+
+    void RemoveAll(CResource* pResource)
+    {
+        std::vector<T> elementsToRemove;
```
What the purpose of using an additional vector? Why don't simple remove elements in the place?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsBaseManager.h:
```diff
@@ -0,0 +1,65 @@
+            ListRemove(m_elementsList, it);
+    }
+
+    void Add(T pElement) { m_elementsList.push_back(pElement); }
```
Wouldn't be better to pass pElement by reference or even using move semantics?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsBaseManager.h:
```diff
@@ -0,0 +1,65 @@
+    void Add(T pElement) { m_elementsList.push_back(pElement); }
+
+    unsigned long Count() const { return m_elementsList.size(); }
+    auto          IterBegin() { return m_elementsList.begin(); }
```
Iterators are so past tense. I thing it wound be better to return const reference to a vector and use within a range-based loop.

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsBoxShape.h:
```diff
@@ -0,0 +1,25 @@
+    CLuaPhysicsBoxShape(CVector half);
+
+    virtual ePhysicsElementType GetType() const { return ePhysicsElementType::BoxShape; }
+    btBoxShape*                 GetBtShape() const { return (btBoxShape*)InternalGetBtShape(); }
```
Why are you using C-style cast?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsElement.h:
```diff
@@ -0,0 +1,39 @@
+
+    virtual ePhysicsElementType GetType() const { return ePhysicsElementType::Shape; }
+
+    const char* GetName() { return EnumToString(GetType()); }
```
Getters should be marked as constant

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsElement.h:
```diff
@@ -0,0 +1,39 @@
+
+    const char* GetName() { return EnumToString(GetType()); }
+
+    void SetOwnedResource(CResource* pResource) { m_pOwnedResource = pResource; }
```
It is recommended to place setter's implementation within source filles instead of headers.

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+
+#include <StdInc.h>
+
+CLuaPhysicsRigidBody::CLuaPhysicsRigidBody(CLuaPhysicsShape* pPhysicsShape, float fMass, CVector vecLocalInertia, CVector vecCenterOfMass)
```
You should probably replace CVector with const CVector&

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+    pPhysicsShape->AddRigidBody(this);
+}
+
+void CLuaPhysicsRigidBody::SetPosition(CVector vecPosition)
```
You should probably replace CVector with const CVector&

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+
+void CLuaPhysicsRigidBody::SetPosition(CVector vecPosition)
+{
+    btTransform& transform = m_pRigidBodyProxy->getWorldTransform();
```
It seems like btTransform& transform is an immutable variable. It's better to make it const.

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+    return CPhysicsSharedLogic::GetPosition(transform);
+}
+
+void CLuaPhysicsRigidBody::SetRotation(CVector vecRotation)
```
You should probably replace CVector with const CVector&

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+    return CPhysicsSharedLogic::GetRotation(transform);
+}
+
+void CLuaPhysicsRigidBody::SetVelocity(CVector vecVelocity)
```
You should probably replace CVector with const CVector&

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.cpp:
```diff
@@ -0,0 +1,84 @@
+    return CPhysicsSharedLogic::ConvertVector(m_pRigidBodyProxy->getLinearVelocity());
+}
+
+void CLuaPhysicsRigidBody::SetAngularVelocity(CVector vecVelocity)
```
You should probably replace CVector with const CVector&

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsRigidBody.h:
```diff
@@ -0,0 +1,49 @@
+    CPhysicsRigidBodyProxy* GetCollisionObject() const { return m_pRigidBodyProxy.get(); }
+
+    // Set's whatever rigid body is added to physics world
+    void SetEnabled(bool bEnabled) { m_pRigidBodyProxy->SetEnabled(bEnabled); }
```
It is recommended to place setter's implementation within source files instead of headers.

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    return ConvertVector(transform.getOrigin());
+}
+
+void CPhysicsSharedLogic::SetPosition(btTransform& transform, const CVector vecPosition)
```
Why don't you pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    transform.setOrigin(ConvertVector(vecPosition));
+}
+
+void CPhysicsSharedLogic::SetRotation(btCollisionObject* pCollisionObject, const CVector vecRotation)
```
Why don't you pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    pCollisionObject->setWorldTransform(transform);
+}
+
+void CPhysicsSharedLogic::SetPosition(btCollisionObject* pCollisionObject, const CVector vecPosition)
```
Why don't you pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+
+CVector CPhysicsSharedLogic::GetRotation(btCollisionObject* pCollisionObject)
+{
+    btTransform& transform = pCollisionObject->getWorldTransform();
```
It seems like btTransform& transform is an immutable variable. It's better to make it const.

tederis replied:
Or even better replace it with **return GetRotation(pCollisionObject->getWorldTransform());**

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+
+CVector CPhysicsSharedLogic::GetPosition(btCollisionObject* pCollisionObject)
+{
+    btTransform& transform = pCollisionObject->getWorldTransform();
```
It seems like btTransform& transform is an immutable variable. It's better to make it const. Or even better replace it with return GetPosition(pCollisionObject->getWorldTransform());

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    return GetPosition(transform);
+}
+
+btBoxShape* CPhysicsSharedLogic::CreateBox(const CVector half)
```
Why don't you pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    return new btBoxShape(ConvertVector(half));
+}
+
+std::unique_ptr<btRigidBody> CPhysicsSharedLogic::CreateRigidBody(btCollisionShape* pBtShape, float fMass, CVector vecLocalInertia, CVector vecCenterOfMass)
```
Why don't you pass CVector by a const reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    z *= 180.0 / PI;
+
+#pragma warning(push)
+#pragma warning(disable : 4244)
```
It seems weird. Is it really needed to disable warning here?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+#pragma warning(pop)
+}
+
+void CPhysicsSharedLogic::CheckMaximumPrimitiveSize(const CVector vector)
```
Why don't pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+        throw std::invalid_argument(SString("Maximum x,y,z must be equal or smaller than %.02f units", BulletPhysics::Limits::MaximumPrimitiveSize).c_str());
+}
+
+void CPhysicsSharedLogic::CheckMinimumPrimitiveSize(const CVector vector)
```
Why don't pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsSharedLogic.cpp:
```diff
@@ -0,0 +1,194 @@
+    CheckMaximumPrimitiveSize(value);
+}
+
+void CPhysicsSharedLogic::CheckPrimitiveSize(const CVector vector)
```
Why don't pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsStaticCollision.cpp:
```diff
@@ -0,0 +1,56 @@
+    GetCollisionObject()->setWorldTransform(transform);
+}
+
+void CLuaPhysicsStaticCollision::SetRotation(CVector vecRotation)
```
Why don't pass CVector by a reference?

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsStaticCollision.cpp:
```diff
@@ -0,0 +1,56 @@
+
+const CVector CLuaPhysicsStaticCollision::GetPosition() const
+{
+    btTransform transform;
```
It would be better to write it in just one line.

tederis replied:
Or even better **return CPhysicsSharedLogic::GetPosition(GetCollisionObject()->getWorldTransform());**

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsStaticCollision.cpp:
```diff
@@ -0,0 +1,56 @@
+{
+    btTransform transform;
+    transform = GetCollisionObject()->getWorldTransform();
+    return CPhysicsSharedLogic::GetRotation(transform);
```
It's better to return CPhysicsSharedLogic::GetRotation(GetCollisionObject()->getWorldTransform());

tederis commented on Shared/mods/deathmatch/logic/physics/CLuaPhysicsStaticCollision.h:
```diff
@@ -0,0 +1,40 @@
+    virtual ePhysicsElementType   GetType() const { return ePhysicsElementType::StaticCollision; }
+
+    // Set's whatever static collision is added to physics world
+    void SetEnabled(bool bEnabled) { m_btCollisionObject->SetEnabled(bEnabled); }
```
It's better to move setter implementation within a source file.

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void CPhysicsDebugDrawer::DrawLineInternal(const btVector3& from, const btVector3& to, const btVector3& lineColor)
+{
+    const CVector& vecFrom = CPhysicsSharedLogic::ConvertVector(from);
```
There is no need for a reference

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.cpp:
```diff
@@ -0,0 +1,61 @@
+    const float    distance = DistanceBetweenPoints3D((vecFrom + vecTo) / 2, m_vecCameraPosition);
+    if (distance < m_fDrawDistance)
+    {
+        float alpha = 255 - distance / m_fDrawDistance * 255;
```
It think it's better to use floats instead of integers in this place.

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.h:
```diff
@@ -0,0 +1,59 @@
+class CPhysicsDebugLine
+{
+public:
+    CPhysicsDebugLine(const CVector& from, const CVector& to, const SColorARGB& color) : from(from), to(to), color(color) {}
```
You are trying to copy values to variables with the same name

CrosRoad95 replied:
![image](https://user-images.githubusercontent.com/22455534/189690162-5d6ce750-5592-471c-8064-16f39ee04894.png)
![image](https://user-images.githubusercontent.com/22455534/189690180-e6721ccd-46e9-46f5-9be8-dd640ccc7aaa.png)
nope, c++ can initialize them appropriately

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsDebugDrawer.h:
```diff
@@ -0,0 +1,59 @@
+    CVector         m_vecCameraPosition;
+
+public:
+    void  SetLineWidth(float fWidth) { m_fLineWidth = fWidth; }
```
It's better to move setter implementation within a source file.

tederis replied:
This also applies to getters below.

tederis commented on Shared/mods/deathmatch/logic/physics/CPhysicsRigidBodyProxy.h:
```diff
@@ -0,0 +1,47 @@
+
+    CPhysicsRigidBodyProxy(const btRigidBodyConstructionInfo& constructionInfo) : btRigidBody(constructionInfo){};
+
+    static std::unique_ptr<CPhysicsRigidBodyProxy> New(CLuaPhysicsShape* pPhysicsShape, const float fMass, CVector& vecLocalInertia, CVector& vecCenterOfMass,
```
It would be preferable to make CVector& vecLocalInertia and CVector& vecCenterOfMass const.

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

# Conflicts:
#	Server/mods/deathmatch/StdInc.h
#	Server/mods/deathmatch/logic/lua/LuaCommon.h
#	Shared/data/MTA San Andreas/MTA/locale/en_US/client.pot

CrosRoad95 committed:
update usings ( not working )

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
fix

CrosRoad95 committed:
fix along tederis suggestions

CrosRoad95 committed:
fix

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

botder committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

# Conflicts:
#	Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.h

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 commented:
Updated branch, added "setElementRotation" test, pr keep waiting for review! :) 
![image](https://user-images.githubusercontent.com/22455534/203850375-a0ce1a92-7bc4-4188-a7bd-360709ab1e6e.png)

CrosRoad95 committed:
update bp to 3.24

CrosRoad95 committed:
use "new" style of lua function declaration

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

TheNormalnij commented:
Can you show real usage for this feature?

CrosRoad95 commented:
> Can you show real usage for this feature?

don't have any more realistic examples yet, but i'm since few days working on feature branch to be able later to just create prs out of it.

Example real examples that don't need to completly revamp gtasa collisions ere for example:
- ragdolls
- Einheit-101 said his gamemode lacking of precise physics tests using rays
- fps/dayz servers as above, you can make very precise players hitboxes, make for example shotgun shoots multiple bullets that drops over time
- there is a rocket league like server, they made ball physics themself. Seems to work but how more convinient it is to just spawn a ball? you can attach bp collision to vehicles and noone notice there's something wrong
- survival servers: if you drop an item currently it stay in place, what if an item if affected by physics? if you drop your item near cliff so item can drop into the water?

following examples are not that hard to implement, maybe i will try make some of them as a demo using feature branch i'm working on

Dutchman101 commented:
Today there was a discussion on dev discord, that i think is valuable for putting a few things into perspective, should CrosRoad95's large PR's like this and V8 engine turn out to not get merged. I can very well relate to devs that are hestitant, and dont want to, as they say, burn their fingers on it

The discussion: [https://imgur.com/a/seD0hqQ](https://imgur.com/a/seD0hqQ)

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

patrikjuvonen committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
Merge remote-tracking branch 'origin/master' into BulletPhysics3D

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D

CrosRoad95 committed:
Merge branch 'master' into BulletPhysics3D
